text/0087-data-types-extensibility.md

```diff
@@ -0,0 +1,130 @@
+
+```rust
+pub enum Block {
+    V1(BlockV1),
+    V2(BlockV2),
+}
+
+pub struct BlockV1 {
+    pub(super) hash: BlockHash,
+    pub(super) header: BlockHeader,
+    pub(super) body: BlockBodyV1,
+}
+
+pub struct BlockV2 {
+    pub(super) hash: BlockHash,
+    pub(super) header: BlockHeader,
+    pub(super) body: BlockBodyV2,
+    pub(super) new_member: bool,
+```
+
+Each specific variant is (de)serializable and there is also a (de)serialization implemented for the top level `enum` wrapper which uses the predefined `TAG`s, for example:
+
+```rust
+    fn to_bytes(&self) -> Result<Vec<u8>, bytesrepr::Error> {
+        let mut buffer = bytesrepr::allocate_buffer(self)?;
+        match self {
+            Block::V1(v1) => {
+                buffer.insert(0, BLOCK_V1_TAG);
+                buffer.extend(v1.to_bytes()?);
+            }
+            Block::V2(v2) => {
+                buffer.insert(0, BLOCK_V2_TAG);
+                buffer.extend(v2.to_bytes()?);
+            }
+        }
+        Ok(buffer)
+    }
+```
+
+Components that never need to touch the legacy types should directly use the most recent version of the datatype. For example, going forward, we never need to gossip "legacy" blocks during the normal chain operation, therefore the `Gossiper` should work directly with `BlockV2`.
+
+The `Fetcher` on the other hand must be able to deal with all variants of the datatype, because it must be able to fetch and understand all historical blocks up to genesis.
+
+## Storage
+
+Storage is the main component that should understand and work with the versioned types. For each data type that needs to be extensible we create a separate DB which will be used to store the top-level `enum` wrapper. The legacy DB is left intact and it's going to be accessed only when old-enough data is requested. Therefore, upon the request to the storage being sent, storage will first look for a particular item in the new database and only if it is not found it'll reach out to the legacy DB and try to obtain data from there.
+
+This is a pseudocode of how a block body retrieval function could be written:
+```rust
+fn get_body_for_block_header(hash, db_current, db_legacy)
+) -> Option<BlockBody> {
+    let body = get_value(hash, db_current);
+    if body.is_none() {
+        get_value(hash, db_legacy)
+    }
+    else {
+        body
+    }
+}
+```
+There is no noticeable performance degradation expected because most of the requests will be saturated from the new DB and we'll fall back to the legacy one only for limited use cases (like historical sync, which in most cases is a one-time operation).
+
+Upon storage initialization, indices are built based on the content of both DBs. Similarly, all utility functions, like `value_exists()` will be updated to check both databases.
+
+# Alternatives
```

Comment:
Since the title of the CEP has heading 1 I think that this section should have heading 2 `##` and all of its subsections should have heading 3

Reply:
Headings fixed in https://github.com/casper-network/ceps/pull/87/commits/a1e164f0a5f7c0ad1a170f6f112ba5759b6eb6de